app/scripts/controllers/network/network-controller.test.ts

```diff
@@ -7888,72 +8040,88 @@ async function waitForPublishedEvents({
   beforeResolving = async () => {
     // do nothing
   },
-}) {
-  const promiseForEventPayloads = new Promise((resolve, reject) => {
-    // We need to declare this variable first, then assign it later, so that
-    // ESLint won't complain that resetTimer is referring to this variable
-    // before it's declared. And we need to use let so that we can assign it
-    // below.
-    /* eslint-disable-next-line prefer-const */
-    let eventListener;
-    let timer;
-    const allEventPayloads = [];
-    const interestingEventPayloads = [];
-    let alreadyEnded = false;
-
-    const end = () => {
-      if (!alreadyEnded) {
-        alreadyEnded = true;
-        messenger.unsubscribe(eventType.toString(), eventListener);
-        Promise.resolve(beforeResolving()).then(() => {
-          if (interestingEventPayloads.length === expectedNumberOfEvents) {
-            resolve(interestingEventPayloads);
-          } else {
-            // Using a string instead of an Error leads to better backtraces.
-            /* eslint-disable-next-line prefer-promise-reject-errors */
-            reject(
-              `Expected to receive ${expectedNumberOfEvents} ${eventType} event(s), but received ${
-                interestingEventPayloads.length
-              } after ${timeBeforeAssumingNoMoreEvents}ms.\n\nAll payloads:\n\n${inspect(
-                allEventPayloads,
-                { depth: null },
-              )}`,
-            );
-          }
-        });
-      }
-    };
+}: {
+  messenger: ControllerMessenger<never, NetworkControllerEvent>;
+  eventType: E['type'];
+  count?: number;
+  filter?: (payload: E['payload']) => boolean;
+  wait?: number;
+  operation?: () => void | Promise<void>;
+  beforeResolving?: () => void | Promise<void>;
+}): Promise<E['payload'][]> {
+  const promiseForEventPayloads = new Promise<E['payload'][]>(
+    (resolve, reject) => {
+      // Everything depends on `eventListener` circularly, so we need to declare
```

Comment:
Nit: This could be simplifed by using a function declaration for `end`, `stopTimer`, and `resetTimer`. Function declarations make circular references easier to handle due to function hoisting; the declaration is hoisted to the top of the scope, so they can be referenced before the declaration.

That lets use `const` when defining the event listener, and saves us from needing this comment to explain the awkward order of operations.